fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+        """
+        if self._apply_layerwise_scaling:
+            for elt in self.layer_info:
+                if not elt.found_inf_or_nan:
+                    for param_name, param in elt.layer_type.named_parameters():
+                        if hasattr(param, "grad"):
+                            logging.debug(
+                                "%s scaling down %s by %s" % (elt.layer_name, param_name, 1.0 / elt.scaling_factor)
+                            )
+                            param.grad = torch.mul(param.grad, 1.0 / elt.scaling_factor)
+
+            while len(self._handles) > 0:
+                elt = self._handles.pop()
+                elt.remove()
+
+    def _check_for_inf_or_nan(self) -> None:
+        """
+        For each layer, check if any of the parameters with a gradient attribute
+        contain an inf/nan. If any of the parameters' gradient contain an inf/nan,
+        then that layers' found_inf_or_nan attribute is set to True and all
+        remaining parameters for that layer are skipped.
+        """
+        for elt in self.layer_info:
+            elt.found_inf_or_nan = False
+            for _, param in elt.layer_type.named_parameters():
+                if hasattr(param, "grad") and param.grad is not None:
+                    if torch.isinf(param.grad).any().item() is True or torch.isnan(param.grad).any().item() is True:  # type: ignore
+                        elt.found_inf_or_nan = True
+                        break  # skip all remaining named parameters
+
+    def step(self, optimizer) -> None:  # type: ignore
+        """
+        If there are NO inf/nan in the gradients' of ALL layers, then optimizer
+        takes a step, otherwise not. Update the scaling factor for each layer.
+        """
+        # using layerwise gradient scaling
+        if self._apply_layerwise_scaling:
+            self._check_for_inf_or_nan()
+            inf_nan_found = any(elt.found_inf_or_nan for elt in self.layer_info)
+
+            if not inf_nan_found:
+                optimizer.step()
+            self._update_scale()
+        # not using layerwise gradient scaling
+        else:
+            optimizer.step()
+
+    def _update_scale(self) -> None:
+        """
+        For each layer, if an inf/nan is found, then multiply the scaling factor
+        of that layer by the backoff factor and set the growth tracker of that
+        layer to 0. Else, increment the growth tracker of the layer. If growth
+        tracker equals the growth interval, then multiply the scaling factor of
+        the layer by the growth factor and reset the layers' growth tracker to 0.
+        Finally, clip the scaling factor to the range
+        [self.min_scaling_factor, self.max_scaling_factor]. The min/max scaling
+        factor values are user configurable.
+        """
+        if self._apply_layerwise_scaling:
+            for layer in self.layer_info:
+                if not layer.found_inf_or_nan:
```

Comment:
If you have if/else constructs its generally easier to read if (true): else(false) instead of having `if not (true)`.

Reply:
agreed and changed.